scripts/gen.py

```diff
@@ -1033,27 +1034,37 @@ def generate_registrations(fgens, overrides):
   return aten_code + '\n}\n' + autogradxla_code + '\n}\n', overridden
 
 
-# XLA is required to register kernel to leaf nodes when there's no math
-# kernel provided in PyTorch core.
-# For other non-leaf nodes, PyTorch covers both forward and backward for
-# them. But XLA can still optionally override them as necessary.
-def requires_registration(fgen):
-  return (fgen.leaf and not fgen.math
-         ) or fgen.mapsig in _FN_AUTOGRAD_XLA or fgen.func in _FN_AUTOGRAD_XLA
-
-
-def generate_functions(fgens):
+# For an op that requires_lowering=True:
+#   - If XLA has a lowering in AtenXlaType, we use register that kernel to XLA dispatch key.
+#   - If XLA doesn't have a lowering in AtenXlaType, we generate one in AtenXlaTypeDefault
+#     and register that kernel to XLA dispatch key.
+# For an op that requires_lowering=False, has_xla_lowering=True:
+#   - xla lowering in AtenXlaType will be registerd to XLA dispatch key and used.
+#     (note this is new  since it was impossible to override composite op before)
+# For an op that has_autogradxla=True:
+#   - the kernel must have both forward and backward implemented and it's registered to
+#     AutogradXLA dispatch key.
+#     Backends should only use this when Autograd kernel in PyTorch codebase doesn't fit.
+#     E.g max_pool2d which enforce materializing indices for backward pass to use.
+def requires_registration(fgen, overrides):
+  requires_lowering = (fgen.dispatch and not fgen.math)
```

Comment:
I am not sure we need parenthesis here.

Nit: The code below could be rewritten to allow `OR` shortcutting (starting from the more unlikely condition), but naming the single conditions clearly has its value.

Reply:
Yep I rewrite the shortcut so that it's has names and easier to match back to the note :D